src/device/asgard_glkronmult_bcpu.cpp

```diff
@@ -0,0 +1,671 @@
+      ASGARD_PRAGMA_OMP_SIMD(collapse(7))
+      for (int w = 0; w < n; w++)
+        for (int j = 0; j < n; j++)
+          for (int s = 0; s < n; s++)
+            for (int v = 0; v < n; v++)
+              for (int p = 0; p < n; p++)
+                for (int l = 0; l < n; l++)
+                  for (int k = 0; k < n; k++)
+                    y[ipow<n, 5>() * w + ipow<n, 4>() * s + ipow<n, 3>() * v + n * n * p + n * l + k] +=
+                        x[ipow<n, 5>() * w + ipow<n, 4>() * j + ipow<n, 3>() * v + n * n * p + n * l + k] * A[j * n + s];
+    }
+    else
+    {
+      ASGARD_PRAGMA_OMP_SIMD(collapse(7))
+      for (int j = 0; j < n; j++)
+        for (int s = 0; s < n; s++)
+          for (int w = 0; w < n; w++)
+            for (int v = 0; v < n; v++)
+              for (int p = 0; p < n; p++)
+                for (int l = 0; l < n; l++)
+                  for (int k = 0; k < n; k++)
+                    y[ipow<n, 5>() * s + ipow<n, 4>() * w + ipow<n, 3>() * v + n * n * p + n * l + k] +=
+                        x[ipow<n, 5>() * j + ipow<n, 4>() * w + ipow<n, 3>() * v + n * n * p + n * l + k] *
+                        A[j * n + s];
+    }
+  }
+}
+
+int64_t number_of_blocks_;
+
+template<typename precision, permutes::matrix_fill fill, int num_dimensions, int dim, int n>
+void global_cpu(int64_t block_size,
+                vector2d<int> const &ilist, dimension_sort const &dsort,
+                connect_1d const &conn,
+                std::vector<precision> const &vals,
+                precision const x[], precision y[],
+                std::vector<std::vector<int64_t>> &row_wspace)
+{
+  //std::cout << " ------------------------------------------------------ \n";
+  constexpr int n2 = n * n;
+
+  int const num_vecs = dsort.num_vecs(dim);
+
+#ifdef _OPENMP
+  int const max_threads = omp_get_max_threads();
+#else
+  int max_threads = 1;
+#endif
+
+  if (static_cast<int>(row_wspace.size()) < max_threads)
+    row_wspace.resize(max_threads);
+
+  int threadid = 0;
+#pragma omp parallel
+  {
+    int tid;
+#pragma omp critical
+    tid = threadid++;
+
+    // xidx holds indexes for the entries of the current
+    // sparse row that are present in the current ilist
+    std::vector<int64_t> &xidx = row_wspace[tid];
+    if (static_cast<int>(xidx.size()) < conn.num_rows())
+      xidx.resize(conn.num_rows(), -1);
+
+#pragma omp for schedule(dynamic)
+    for (int vec_id = 0; vec_id < num_vecs; vec_id++)
+    {
+      int const vec_begin = dsort.vec_begin(dim, vec_id);
+      int const vec_end   = dsort.vec_end(dim, vec_id);
+      // map the indexes of present entries
+      for (int j = vec_begin; j < vec_end; j++)
+        xidx[dsort(ilist, dim, j)] = dsort.map(dim, j) * block_size;
+
+      // matrix-vector product using xidx as a row
+      for (int rj = vec_begin; rj < vec_end; rj++)
+      {
+        // row in the 1d pattern
+        int const row = dsort(ilist, dim, rj);
+
+        precision *const local_y = &y[xidx[row]];
```

Comment:
```suggestion
        precision *const local_y = y + xidx[row];
```
or use [std::next](https://en.cppreference.com/w/cpp/iterator/next)